citar-file.el

```diff
@@ -106,36 +115,42 @@ Example: ':/path/to/test.pdf:PDF'."
          (mapcar (apply-partially #'expand-file-name fn) dirs)))
      parts)))
 
-(defun citar-file--possible-names (key dirs extensions &optional entry)
+(defun citar-file--possible-names (key dirs extensions &optional entry find-additional)
   "Possible names for files correponding to KEY, ENTRY with EXTENSIONS in DIRS."
   (cl-flet ((possible-file-names-with-extension
              (extension)
              (seq-map
               (lambda (directory)
-                (expand-file-name
-                 (concat key "." extension) directory))
+                (if find-additional                    
+                    (directory-files directory t
+                                     (concat "^" (regexp-quote key)
+                                             ".*\\("
+                                             extension
+                                             "\\)$"))
+                  (list (expand-file-name
+                         (concat key "." extension) directory))))
               dirs)))
-    (let* ((results-key (seq-mapcat
-                         #'possible-file-names-with-extension
-                         extensions))
+    (let* ((results-key (apply #'append (seq-mapcat
+                                         #'possible-file-names-with-extension
+                                         extensions)))
            (file-field (citar-get-value
                         citar-file-variable entry))
            (results-file
-           (when file-field
-             (seq-mapcat
-              (lambda (func)
-                (funcall
-                 func
-                 ;; Make sure this arg is non-nil.
-                 (or dirs "/")
-                 file-field))
-              citar-file-parser-functions))))
+            (when file-field
+              (seq-mapcat
+               (lambda (func)
+                 (funcall
+                  func
+                  ;; Make sure this arg is non-nil.
+                  (or dirs "/")
+                  file-field))
+               citar-file-parser-functions))))
       (delete-dups (append results-key results-file)))))
 
 (defun citar-file--files-for-entry (key entry dirs extensions)
     "Find files related to KEY, ENTRY in DIRS with extension in EXTENSIONS."
     (seq-filter #'file-exists-p
-                (citar-file--possible-names key dirs extensions entry)))
+                (citar-file--possible-names key dirs extensions entry citar-file-find-additional-files)))
```

Comment:
If the defcustom is set to `t`, will this not run under `[citar--format-candidates](https://github.com/bdarcus/citar/blob/6e73acd559a1d60c74af58e1794f4f1e389e5df9/citar.el#L443)`?

See in particular:

https://github.com/bdarcus/citar/blob/6e73acd559a1d60c74af58e1794f4f1e389e5df9/citar.el#L457-L462

I guess we should clarify what the defcustom does.

My impulse is to say only turns on file finding for the open commands, but is not used in candidate generation.

I suppose ideally we could do some profiling to figure out how significant the performance concern is. The test would be a large file (say 2000+ entries), and multiple directories.